clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -633,11 +716,22 @@ public synchronized void handleFailedBatch(ProducerBatch batch, RuntimeException
             // Reset the producerId since we have hit an irrecoverable exception and cannot make any guarantees
             // about the previously committed message. Note that this will discard the producer id and sequence
             // numbers for all existing partitions.
-            resetIdempotentProducerId();
+            requestEpochBumpForPartition(batch.topicPartition);
+        } else if (exception instanceof UnknownProducerIdException) {
+            // If we get an UnknownProducerId for a partition, then the broker has no state for that producer. It
+            // will therefore accept a write with sequence number 0. We reset the sequence number for the partition
+            // here so that the producer can continue after aborting the transaction. Note that if the broker supports
+            // bumping the epoch, we will later reset all sequence numbers after calling InitProducerId
+            resetSequenceForPartition(batch.topicPartition);
```

Comment:
I think this works, just checking my understanding. Suppose we still have inflight requests and we are talking to a broker that doesn't support epoch bumps. When we reset sequence numbers above, we also drop the state of those inflight requests. I think we are implicitly relying on those inflights also returning the same UNKNONW_PRODUCER_ID error. Is that right?

Reply:
That's the assumption I was making, yes. If we have one request in-flight that gets an UNKNOWN_PRODUCER_ID, I wouldn't expect any other in-flights to be able to succeed.